### PR TITLE
Make WebsiteAgent merge `template` with the results of `extract`

### PR DESCRIPTION
**Pull request by knu**
A new extraction option `hidden` is added so that keys with it gets excluded from the final payloads while they can be used in `template`.

### CODE REVIEW

File: app/models/agents/website_agent.rb [hunk 1]
```diff
@@ -159,7 +159,11 @@ class WebsiteAgent < Agent
     end
 
     def event_keys
-      (options['template'].presence || options['extract']).try(:keys)
+      extract = options['extract'] or return nil
+
+      extract.each_with_object([]) { |(key, value), keys|
+        keys << key unless boolify(value['hidden'])
+      } | (options['template'].presence.try!(:keys) || [])
```
Comment (cantino) [inline app/models/agents/website_agent.rb]: What's the difference between `try` and `try!`?

File: app/models/agents/website_agent.rb [hunk 2]
```diff
@@ -116,11 +118,9 @@ class WebsiteAgent < Agent
 
       Set `http_success_codes` to an array of status codes (e.g., `[404, 422]`) to treat HTTP response codes beyond 200 as successes.
 
-      If a `template` option is given, it is used as a Liquid template for each event created by this Agent, instead of directly emitting the results of extraction as events.  In the template, keys of extracted data can be interpolated, and some additional variables are also available as explained in the next section.  For example:
+      If a `template` option is given, its value must be a hash, whose key-value pairs are interpolated after extraction for each iteration and merged with the payload.  In the template, keys of extracted data can be interpolated, and some additional variables are also available as explained in the next section.  For example:
```
Comment (cantino) [inline app/models/agents/website_agent.rb]: I think this is a good change.

File: app/models/agents/website_agent.rb [hunk 3]
```diff
@@ -538,7 +528,12 @@ def extract_each(&block)
           block.call(extraction_details, values)
         end
         log "Values extracted: #{values}"
-        output[name] = values
+        begin
+          output[name] = values
+          output.hidden_keys << name if boolify(extraction_details['hidden'])
+        rescue ArgumentError
```
Comment (cantino) [inline app/models/agents/website_agent.rb]: Maybe a specific error subclass?

**Comment by cantino** [comment]
I like this change

File: app/models/agents/website_agent.rb (hunk 3 above)
Comment (knu) [inline app/models/agents/website_agent.rb]: Good point, done!

**Pull request merged**